pgrx/src/datum/array.rs

```diff
@@ -231,7 +240,18 @@ impl<'mcx, T: UnboxDatum> Array<'mcx, T> {
     ) -> Option<T::As<'arr>> {
         match is_null {
             true => None,
-            false => unsafe { self.slide_impl.bring_it_back_now(self, ptr) },
+            false => {
+                // Ensure we are not attempting to dereference a pointer
+                // outside of the array.
+                debug_assert!(self.is_within_bounds(ptr));
+                // Prevent a datum that begins inside the array but would end
+                // outside the array from being dereferenced.
+                debug_assert!(self.is_within_bounds_inclusive(
+                    ptr.wrapping_add(unsafe { self.slide_impl.hop_size(ptr) })
+                ));
```

Comment:
If we used a fn from the `.offset` family instead of `.wrapping_offset` then it's simply UB and any such check is elided, but `hop_size` does not do this inherently...